neps/search_spaces/yaml_search_space_utils.py

```diff
@@ -82,32 +81,44 @@ def __init__(self, exception):
         super().__init__(self.message)
 
 
-def deduce_and_validate_param_type(
-    name: str, details: dict[str, str | int | float]
+def deduce_type(
+    name: str, details: dict[str, str | int | float] | str | int | float
 ) -> str:
     """
-    Deduces the parameter type from details and validates them.
+    Deduces the parameter type from details.
 
     Args:
         name (str): The name of the parameter.
-        details (dict): A dictionary containing parameter specifications.
+        details (dict | str | int | float): A dictionary containing parameter
+                specifications or a direct value (string, integer, or float).
 
     Returns:
         str: The deduced parameter type ('int', 'float', 'categorical', or 'constant').
 
     Raises:
         TypeError: If the type cannot be deduced or the details don't align with expected
-        constraints.
-    """
-    # Deduce type
-    if "type" in details:
-        param_type = details["type"].lower()
-    else:
-        # Logic to infer type if not explicitly provided
-        param_type = deduce_param_type(name, details)
-
-    # Validate details of a parameter based on (deduced) type
-    validate_param_details(name, param_type, details)
+                constraints.
+        """
+    try:
+        # Deduce type
+        if "type" in details:
+            param_type = details.pop("type").lower()
+        else:
+            # because details could be string
+            if isinstance(details, (str, int, float)):
+                param_type = "const"
+                return param_type
+            else:
+                param_type = deduce_param_type(name, details)
+    except TypeError as e:
+        # because details could be int, float
+        if isinstance(details, (str, int, float)):
+            param_type = "const"
+            return param_type
+        else:
+            raise TypeError(
+                f"Unable to deduce parameter type for '{name}' with details '{details}'.") \
+                from e
```

Comment:
I think it would be much cleaner and easier to understand this by starting with something like:

```python
if isinstance(details, (str, int, float)):
   ... # assume constant
   return something
   
if isinstance(details, dict):
   ... # It's a dict, do what you do
   return something
   
raise TypeError() # Got something we didn't expect to get
```

A lot less try/catches and it's clearer there's a seperate path for each possibility.

```

Reply:
The idea was in the first place to be more efficient because of the less usage of constant parameter but then I got lost in all these cases. I changed it in the new PR